matchzoo/contrib/layers/spatial_gru.py

```diff
@@ -0,0 +1,243 @@
+                 direction: str = 'lr',
+                 **kwargs):
+        """:class:`SpatialGRU` constructor."""
+        super().__init__(**kwargs)
+        self._units = units
+        self._normalize = normalize
+        self._init_diag = init_diag
+        self._activation = activations.get(activation)
+        self._recurrent_activation = activations.get(recurrent_activation)
+        self._use_bias = use_bias
+
+        self._kernel_initializer = initializers.get(kernel_initializer)
+        self._recurrent_initializer = initializers.get(recurrent_initializer)
+        self._bias_initializer = initializers.get(bias_initializer)
+        self._direction = direction
+
+    def build(self, input_shape: typing.Any):
+        """
+        Build the layer.
+
+        :param input_shape: the shapes of the input tensors.
+        """
+        # input_shape: (batch_size, channel, text1_maxlen, text2_maxlen)
+        self._batch_size = input_shape[0]
+        self._channel = input_shape[1]
+        self._input_dim = self._channel + 3 * self._units
+
+        self._text1_maxlen = input_shape[2]
+        self._text2_maxlen = input_shape[3]
+        self._recurrent_step = self._text1_maxlen * self._text2_maxlen
+
+        self._W = self.add_weight(name='W',
+                                  shape=(self._input_dim, self._units * 7),
+                                  initializer=self._kernel_initializer,
+                                  trainable=True)
+
+        self._U = self.add_weight(name='U',
+                                  shape=(self._units * 3, self._units),
+                                  initializer=self._recurrent_initializer,
+                                  trainable=True)
+
+        self._bias = self.add_weight(name='bias',
+                                     shape=(self._units * 8,),
+                                     initializer='zeros',
+                                     trainable=True)
+
+        # w_rl, w_rt, w_rd
+        self._wr = self._W[:, :self._units * 3]
+        # b_rl, b_rt, b_rd
+        self._br = self._bias[:self._units * 3]
+        # w_zi, w_zl, w_zt, w_zd
+        self._wz = self._W[:, self._units * 3: self._units * 7]
+        # b_zi, b_zl, b_zt, b_zd
+        self._bz = self._bias[self._units * 3: self._units * 7]
+        self._w_ij = self.add_weight(name='Wij',
+                                     shape=(self._channel, self._units),
+                                     initializer=self._recurrent_initializer,
+                                     trainable=True)
+        self._b_ij = self._bias[self._units * 7:]
+        super(SpatialGRU, self).build(input_shape)
+
+    def softmax_by_row(self, z: typing.Any) -> tuple:
+        """Conduct softmax on each dimension across the four gates."""
+
+        z_transform = Permute((2, 1))(Reshape((4, self._units))(z))
+        for i in range(0, self._units):
+            begin1 = [0, i, 0]
+            size = [-1, 1, -1]
```

Comment:
why define the same list in a loop? basically you're creating `size` in each interation.

Reply:
`size` has been moved out of the loop.